README.md

```diff
@@ -75,6 +75,73 @@ The PMEM-CSI driver supports running in different modes, which can be controlled
 
 * **_Unified_** mode is intended to run the driver in a single host, mostly for testing the driver in a non-clustered environment.
 
+### Volume Persistency
+
+Normally, volumes are provided by a storage backend that is independent
+of a particular node. When a node goes offline, the volume can be
+mounted elsewhere. With PMEM-CSI, volumes are *local* and thus can
+only be used on the node where they were created.
+
+This limitation needs to be considered when designing and deploying
+applications that are to use local storage:
+
+* Persistent Volumes  
+A volume gets created independently of the application, on some node
+where there is enough free space. Applications using such a volume are
+then forced to run on that node and cannot run when the node is
+down. Data is retained until the volume gets deleted.
+* Ephemeral Volumes  
+An application is assigned to a node and when it starts to run, a new
+volume is created for it on that node. When the application stops, the
+volume is deleted.
+* Cache Volumes  
```

Comment:
@pohly can you please help me in depict this usage on Kuberentes. Applications in the sense stateful/replica sets or might be two different applications? How does PV -> PmemVolume relation 1-1 or 1-many.

Reply:
So different Applications share the same PV.

Just to make sure if we are on same lines on how does this node set are requested by Pod, does my https://github.com/kubernetes-csi/external-provisioner/issues/221#issuecomment-460552787 address this.

Reply:
No, I don't think that comment captures the idea. The choice between "cache volume" and "normal persistent volume" is something that the creator of the PVC must explicitly make by setting a driver-specific parameter. It is not something that pmem-csi deduces from the topology fields, and it is independent of the pods that then end up using the PV.

Reply:
To phrase it differently, persistent and cache volumes are not "requested by a pod". They are requested by an admin by creating a PVC.

The only volumes that can be said to be "requested by a pod" are the ephemeral volumes, because only those are tied to a specific pod.

Reply:
> No, I don't think that comment captures the idea. The choice between "cache volume" and "normal persistent volume" is something that the creator of the PVC must explicitly make by setting a driver-specific parameter

driver-specific parameters you mean StorageClass or PVC? PVC cannot have parameters i guess(i might be wrong) other than Selector. Can't we us SotrageClass.AllowedTopologies to limit the nodes?

Reply:
PVC is more close to User/Pod, StorageClass is for cluster admins i feel.

Reply:
I think the persistency type ("persistent" vs. "ephemeral" vs. "cache") belongs into the StorageClass as a parameter. That's something that the admin needs to make available, normal users usually don't have the rights to create cluster-level, non-namespaced entities like a StorageClass.

We could use use StorageClass.AllowedTopologies. That is converted for the CSI driver by external-provisioner, isn't it?

But my gut feeling is that the choice of nodes needs to be closer to the app deployer, i.e. the non-privileged user, which means offering a parameter in the PVC. But as you said, the ability to influence dynamic provisioning via a PVC are limited. There is PVC.Resources.Requests, but while resource names are strings, the values have to be a number.

I thought we could use the same concept as ceph (let users create secrets where the content of the secret includes parameters for provisioning), but looking closer at https://kubernetes-csi.github.io/docs/Usage.html#csi-provisioner-parameters I only find `${pv.name}` listed as valid template for provisioning. I'm unsure how to use that. I'll do some more research on this.


. This works by having templates in the storage class (:
```

Reply:
> I think the persistency type ("persistent" vs. "ephemeral" vs. "cache") belongs into the StorageClass as a parameter. That's something that the admin needs to make available, normal users usually don't have the rights to create cluster-level, non-namespaced entities like a StorageClass.
> 
I agree, though myself not clear how does persistency type really effect the driver behavior. Atleast i could think of knowing that this volume is used as "cache" at creation time effect the driver, but not sure about "persistent" vs. "ephemeral". My feeling is those are differentiated by the way the volume get deleted. But you can ignore me as long as your are clear with this :)

> We could use use StorageClass.AllowedTopologies. That is converted for the CSI driver by external-provisioner, isn't it?
> 
Yes, right. Thats the reason it attracted me.

> But my gut feeling is that the choice of nodes needs to be closer to the app deployer, i.e. the non-privileged user, which means offering a parameter in the PVC. But as you said, the ability to influence dynamic provisioning via a PVC are limited. There is PVC.Resources.Requests, but while resource names are strings, the values have to be a number.
> 
There is [PVC.Selector](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#selector) can be used by Pod/PVC to select the volumes only match with volume labels.

But currently these selection is not handled by external-provisioner, even if it does in near feature, it would be via CreateVolumeRequest.AccessibleTopology i feel. That is one reason i am negotiating in my issue on external-provisioner.

> I thought we could use the same concept as ceph (let users create secrets where the content of the secret includes parameters for provisioning), but looking closer at https://kubernetes-csi.github.io/docs/Usage.html#csi-provisioner-parameters I only find `${pv.name}` listed as valid template for provisioning. I'm unsure how to use that. I'll do some more research on this.

Thanks for sharing, i will check.